IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/ElementsHandler.cs

```diff
@@ -98,21 +111,52 @@ public bool ContainsText(string text)
         /// </summary>
         /// <param name="predicate">The criteria to attempt to match on.</param>
         /// <returns></returns>
-        public IWebElement GetSingleExistingElement(Func<IWebElement, bool> predicate)
+        public IWebElement GetSingleWebElement(Func<IWebElement, bool> predicate)
+        {
+            IList<IWebElement> elems = GetElements(predicate);
+
+            if (elems.Count != 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(predicate), "The provided predicate did not match exactly one result.");
+            }
+
+            return elems[0];
+        }
+
+        /// <summary>
+        /// Gets all elements found by <see cref="Locator"/>, matching a given predicate.
+        /// </summary>
+        /// <param name="predicate">The function used to filter to one or more IWebElements</param>
+        /// <returns>A list of found IWebElements</returns>
+        /// <exception cref="NoSuchElementException"></exception>
+        /// <exception cref="ArgumentOutOfRangeException"></exception>
+        public IList<IWebElement> GetAllWebElements(Func<IWebElement, bool> predicate)
+        {
+            return GetElements(predicate);
+
+        }
+
+        public IList<IWebElement> GetElements(Func<IWebElement, bool> predicate)
         {
             IWait<IWebDriver> wait = Wait;
             wait.IgnoreExceptionTypes(typeof(NoSuchElementException));
-            return wait.Until(d =>
+
+#pragma warning disable CS8603 // Possible null reference return. Needed for Selenium retry.
```

Comment:
Is it really correct to suppress this warning? If we are going to have cases returning null, shouldn't we change the return type to reflect that?

Reply:
The challenge is Selenium here. The Until() method swallows nulls and takes that as a fail case to retry when the delegate doesn't return bool. If the delegate never returns non-null, then Selenium throws an exception. If, in this scenario, the null was ever _actually_ returned, there would be a bug in Selenium.

Reply:
I think what I would suggest in that case is to just do `return null!;` down on line 156 and leave a comment next to it indicating that selenium treats that as failure.

Reply:
Done and done.